ucp/_libs/ucx_api.pyx

```diff
@@ -446,11 +446,75 @@ cdef class UCXListener(UCXObject):
         return int(<uintptr_t>self._handle)
 
 
+# Counter used as UCXRequest UIDs
+cdef int _ucx_request_counter = 0
+
+cdef class UCXRequest:
+    """Python wrapper of UCX request handle.
+
+    Notice, this class doesn't own the handle and multiple instances of
+    UCXRequest can point to the same underlying UCX handle.
+    Furthermore, UCX can modify/free the UCX handle without notice
+    thus we use `_uid` to make sure the handle hasn't been modified.
+    """
+    cdef:
+        ucp_request *_handle
+        int _uid
+
+    def __init__(self, uintptr_t req_as_int):
+        global _ucx_request_counter
+        cdef ucp_request *req = <ucp_request*>req_as_int
+        assert req != NULL
+        self._handle = req
+
+        cdef dict info = {"status": "pending"}
+        if self._handle.info == NULL:  # First time we are wrapping this UCX request
+            Py_INCREF(info)
+            self._handle.info = <PyObject*> info
+            _ucx_request_counter += 1
+            self._uid = _ucx_request_counter
+            assert self._handle.uid == 0
```

Comment:
According to the class doc, this object doesn't own `self._handle`, is it safe to assume it's not being used elsewhere and cause a failure here?

Reply:
This is what `self._uid` is for. By checking `self._uid == self._handle.uid`, we know that the underlying UCX handle hasn't been modified.

Reply:
I understand that, but is that allowed to happen in some circumstance at runtime? IOW, if that happens is it due to a bug in user code?

Reply:
And by runtime I mean production code, rather than some debug case.

Reply:
Sorry misunderstood your question, yes this can happen. 

Consider a `ucp_tag_send_nb()` call, it can return a UCX handle immediately **and** when the communicate finishes, call `_send_callback(void *request)` where `request` is the same handle .

Reply:
Should we have a class that does own the handle and this refers to? Wondering how we can make sure the handle gets cleaned up correctly.

Reply:
As far as I can see, it is not possible to make UCX create/free a custom request handle. We can set a [`request_init()` and `request_cleanup()`](http://openucx.github.io/ucx/api/latest/html/group___u_c_p___c_o_n_f_i_g.html#ac1f0a25d794111217af5e0e6b9865458) but "_This function will be called only on the very first time a request memory is initialized, and may not be called again if a request is reused_".

Currently, this PR guarantee that `ucp_request_free()` is called _at most_ once and since our send/recv makes sure to free inflight requests, we shouldn't see any leaks.

Reply:
From the last few comments, it seems that `if self._handle.info == NULL:` should normally ensure that we don't go into this block more than once, and if it happens that we do and the assertion fails, something unexpected did happen. Is that correct?

Reply:
Yes, that is correct.

Reply:
Ok, that seems reasonable to me then. I'll defer any final comments to @jakirkham , just in case he sees issues with it.